tee-worker/omni-executor/intent/asset-lock/src/precise.rs

```diff
@@ -0,0 +1,244 @@
+// Copyright 2020-2024 Trust Computing GmbH.
+// This file is part of Litentry.
+//
+// Litentry is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Litentry is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{AssetId, AssetsLock};
+use std::collections::HashMap;
+
+// Manages asset locks per account. Used for precise tracking of currently locked assets.
+pub struct PreciseAssetsLock {
+	locked_assets: HashMap<AssetId, u128>,
+}
+
+impl AssetsLock for PreciseAssetsLock {
+	fn with_lock(
+		asset_id: AssetId,
+		amount_to_lock: u128,
+		available_amount: u128,
+	) -> Result<Self, ()> {
+		if amount_to_lock > available_amount {
+			return Err(());
+		}
+
+		let mut locked_assets = HashMap::new();
+		locked_assets.insert(asset_id, amount_to_lock);
+
+		Ok(Self { locked_assets })
+	}
+
+	fn lock(
+		&mut self,
+		asset_id: AssetId,
+		amount_to_lock: u128,
+		available_amount: u128,
+	) -> Result<(), ()> {
+		if let Some(lock) = self.locked_assets.get_mut(&asset_id) {
+			if *lock + amount_to_lock > available_amount {
+				return Err(());
+			}
+			*lock += amount_to_lock;
+		} else {
+			if amount_to_lock > available_amount {
+				return Err(());
+			}
+			self.locked_assets.insert(asset_id, amount_to_lock);
+		}
+		Ok(())
+	}
+
+	fn release(&mut self, asset_id: AssetId, amount_to_release: u128) -> Result<(), ()> {
+		if let Some(lock) = self.locked_assets.get_mut(&asset_id) {
+			if *lock < amount_to_release {
+				// this is some kind of inconsistency - we should either set lock to 0 and return Ok or return Err
```

Comment:
We could go with less restrictive strategy here and return `Ok` even if not locked.